### PR TITLE
Azure Monitor Exporter - Update logs scope

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
Based on the implementation from [dotnet](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs#L105) and [Application Insights](https://github.com/microsoft/ApplicationInsights-dotnet/blob/ed0015a78a53b3a26e04d18eed442581bc7c571f/LOGGING/src/ILogger/ApplicationInsightsLogger.cs#L186).

```C#
using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.SetMinimumLevel(LogLevel.Trace);
    builder.AddOpenTelemetry(loggerOptions =>
        {
            loggerOptions.IncludeScopes = true;
            loggerOptions.IncludeFormattedMessage = true;
            loggerOptions.AddAzureMonitorLogExporter(exporterOptions =>
            {
                exporterOptions.ConnectionString = "InstrumentationKey=00000000-0000-0000-0000-000000000000;EndpointSuffix=ingestion.azuremonitor.com";
            });
        });
});

var _logger = loggerFactory.CreateLogger<DemoTrace>();

using (_logger.BeginScope("Some name"))
{
    _logger.LogInformation("1");
    using (_logger.BeginScope(42))
    {
        _logger.LogInformation("2");
        using (_logger.BeginScope("Formatted {WithValue}", 12345))
        {
            _logger.LogInformation("3");
            using (_logger.BeginScope(new Dictionary<string, object> { ["ViaDictionary"] = 100 }))
            {
                _logger.LogInformation("Format: {0}", 4);
                _logger.LogInformation("Hello from the Index!");
                _logger.LogDebug("Hello is done");
            }
        }
    }
}
```

```json
{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "1",
		"severityLevel": "Information",
		"properties": {
			"OriginalFormat": "1",
			"Scope": " => Some name"
		},
		"ver": 2
	}
}
}

{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "2",
		"severityLevel": "Information",
		"properties": {
			"OriginalFormat": "2",
			"Scope": " => Some name => 42"
		},
		"ver": 2
	}
}
}

{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "3",
		"severityLevel": "Information",
		"properties": {
			"OriginalFormat": "3",
			"WithValue": "12345",
			"OriginalFormatScope_1": "Formatted 12345",
			"Scope": " => Some name => 42"
		},
		"ver": 2
	}
}
}

{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "Format: 4",
		"severityLevel": "Information",
		"properties": {
			"0": "4",
			"OriginalFormat": "Format: {0}",
			"WithValue": "12345",
			"OriginalFormatScope_1": "Formatted 12345",
			"ViaDictionary": "100",
			"Scope": " => Some name => 42"
		},
		"ver": 2
	}
}
}

{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "Hello from the Index!",
		"severityLevel": "Information",
		"properties": {
			"OriginalFormat": "Hello from the Index!",
			"WithValue": "12345",
			"OriginalFormatScope_1": "Formatted 12345",
			"Formatted": "100",
			"Scope": " => Some name => 42"
		},
		"ver": 2
	}
}
}

{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "Hello is done",
		"severityLevel": "Verbose",
		"properties": {
			"OriginalFormat": "Hello is done",
			"WithValue": "12345",
			"OriginalFormatScope_1": "Formatted 12345",
			"ViaDictionary": "100",
			"Scope": " => Some name => 42"
		},
		"ver": 2
	}
}
}
```

**TODO**

Tests are based on `InMemoryExporter`, it does not expose scope. Will add the test in the follow up PR.

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 1]
```diff
@@ -80,6 +83,45 @@ internal static string GetMessageAndSetProperties(LogRecord logRecord, IDictiona
...
+                        {
+                            properties.Add(scopeItem.Key, Convert.ToString(scopeItem.Value, CultureInfo.InvariantCulture));
+                        }
+                        else
+                        {
+                            AzureMonitorExporterEventSource.Log.Write($"DuplicateScopeItem{EventLevelSuffix.Warning}", $"Found duplicate scope key - {scopeItem.Key}.");
+                        }
+                    }
+                }
+                else
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:112]: is this possible?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 2]
```diff
@@ -80,6 +83,45 @@ internal static string GetMessageAndSetProperties(LogRecord logRecord, IDictiona
...
+
+            void ProcessScope(LogRecordScope scope, IDictionary<string, string> properties)
+            {
+                if (scope.Scope is IEnumerable<KeyValuePair<string, object>> stateDictionary)
+                {
+                    foreach (KeyValuePair<string, object> scopeItem in scope)
+                    {
+                        if (scopeItem.Key == "{OriginalFormat}")
+                        {
+                            properties.Add($"OriginalFormatScope_{originalFormatDepth++}", Convert.ToString(scope.Scope.ToString(), CultureInfo.InvariantCulture));
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:100]: it looks like the depth is incremented *only* for OriginalFormat.. is that intentional?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs (hunk 1 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:112]: Yes. this is possible. From the sample `using (_logger.BeginScope("Some name"))`, calls this `else` part of the code. 

`using (_logger.BeginScope("Formatted {WithValue}", 12345))` invokes the `if` part.

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:112]: That'd be handled by the SDK itself. The `LogRecordScope` has enumerator, so you can simply do foreach(var scope in logRecordScope)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs (hunk 2 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:100]: Yes. It is intentional. There could be multiple original format, this helps track all of it.

```
{
"name": "Message",
...
"data": {
	"baseType": "MessageData",
	"baseData": {
		"message": "3.1",
		"severityLevel": "Verbose",
		"properties": {
			"OriginalFormat": "3.1",
			"WithValue": "12345",
			"OriginalFormatScope_1": "Formatted 12345",
                        "SomeValue": "56789",
                         "OriginalFormatScope_2":"2nd Formatted 56789"
			"ViaDictionary": "100",
			"Scope": " => Some name => 42"
		},
		"ver": 2
	}
}
```

Code

```
using (_logger.BeginScope("Some name"))
{
    _logger.LogInformation("1");
    using (_logger.BeginScope(42))
    {
        _logger.LogInformation("2");
        using (_logger.BeginScope("Formatted {WithValue}", 12345))
        {
            using (_logger.BeginScope("2nd Formatted {SomeValue}", 56789))
            {
                _logger.LogInformation("3.1");
            }
        }
    }
}
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 3]
```diff
@@ -80,6 +83,45 @@ internal static string GetMessageAndSetProperties(LogRecord logRecord, IDictiona
...
+                        }
+                        else if (!properties.TryGetValue(scopeItem.Key, out _))
+                        {
+                            properties.Add(scopeItem.Key, Convert.ToString(scopeItem.Value, CultureInfo.InvariantCulture));
+                        }
+                        else
+                        {
+                            AzureMonitorExporterEventSource.Log.Write($"DuplicateScopeItem{EventLevelSuffix.Warning}", $"Found duplicate scope key - {scopeItem.Key}.");
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:108]: its not likely an actionable item for users, as they may not control scopes. 
An alternative would be to prefix scope depth to make everything unique.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs (hunk 1 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:112]: Correct! I use the enumerator on line 96 from `LogRecordScope`

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs (hunk 3 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:108]: Had the same and remove before last commit. Will add the scope depth.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs (hunk 1 above)
Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:113]: I meant this : https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs#L147-L154

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:117]: If we do only this one (foreach), we will miss out the scope info from `using (_logger.BeginScope("Some name"))`.  I don't think OTLP log exporter tracks it.

This implementation has both foreach and the normal tracking

**Review by vishweshbankwar — APPROVED** [review]

**Pull request merged**